File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    var min = findMin(nums);
    if (min === 0) {
        return binarySearch(nums, target);
    }
    
    if (target < nums[0] && target > nums[nums.length - 1]) {
        return -1;
    }
    
    if (target >= nums[0]) {
        return binarySearch(nums.slice(0, min), target);
    }
    
    var ans = binarySearch(nums.slice(min, nums.length), target);    
    if (ans === -1) return -1;
    return ans + min;
};

var findMin = function(nums) {
    var low = 0;
    var high = nums.length - 1;
    if (nums[low] < nums[high]) {
        return low;
    }
    while (low < high) {
        var mid = Math.floor((low + high) / 2);
        if (nums[mid] > nums[high]) {
            low = mid + 1;
        } else if (nums[mid] < nums[high]) {
            high = mid;
        }
    }
    return low;
}

var binarySearch = function(nums, target) {
    var low = 0;
    var high = nums.length - 1;
    while (low <= high) {
        var mid = Math.floor((low + high) / 2);
        if (nums[mid] > target) {
            high = mid - 1;
        } else if (nums[mid] < target) {
            low = mid + 1;
        } else {
            return mid
        }
    }
    return -1; 
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution(object):
    @staticmethod
    def find_min(nums):
        low = 0
        high = len(nums) - 1
        if nums[low] < nums[high]:
            return low
        while low < high:
            mid = (low + high) // 2
            if nums[mid] < nums[high]:
                high = mid
            elif nums[mid] > nums[high]:
                low = mid + 1 
        return low
    
    @staticmethod
    def binarySearch(nums, target):
        low = 0
        high = len(nums) - 1
        while low <= high:
            mid = (low + high) // 2
            if nums[mid] < target:
                low = mid + 1
            elif nums[mid] > target:
                high = mid - 1 
            else:
                return mid
        return -1
    
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        min_index = Solution.find_min(nums)

        if min_index == 0:
            return Solution.binarySearch(nums, target)
        if target >= nums[0]:
            return Solution.binarySearch(nums[:min_index], target)
        ans = Solution.binarySearch(nums[min_index:], target)
        if ans == -1:
            return -1
        return ans + min_index
